File: languages/python/projects/01-foundations/main.py
def read_positive_int(prompt: str) -> int:
    while True:
        raw = input(prompt).strip()
        try:
            value = int(raw)
        except ValueError:
            print("Please enter a valid integer.")
            continue

        if value <= 0:
            print("Please enter a number greater than zero.")
            continue

        return value


def read_name(index: int) -> str:
    while True:
        name = input(f"Student #{index} name: ").strip()
        if name:
            return name

        print("Name cannot be empty.")


def read_score(name: str) -> int:
    while True:
        raw = input(f"{name}'s score (0-100): ").strip()
        try:
            score = int(raw)
        except ValueError:
            print("Please enter a valid integer score.")
            continue

        if 0 <= score <= 100:
            return score

        print("Score must be between 0 and 100.")


def main() -> None:
    students = []
    total_students = read_positive_int("How many students? ")

    for index in range(1, total_students + 1):
        name = read_name(index)
        score = read_score(name)
        students.append({"name": name, "score": score})

    scores = [student["score"] for student in students]
    average = sum(scores) / len(scores)

    print("Students:")
    for student in students:
        print(f"- {student['name']}: {student['score']}")

    print(f"Average: {average:.2f}")
    print(f"Minimum: {min(scores)}")
    print(f"Maximum: {max(scores)}")


if __name__ == "__main__":
    main()

File: languages/java/projects/01-foundations/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Scanner;

public class Main {
    record Student(String name, int score) {
    }

    static int readPositiveCount(Scanner scanner) {
        while (true) {
            System.out.print("How many students? ");
            String raw = scanner.nextLine().trim();
            try {
                int value = Integer.parseInt(raw);
                if (value > 0) {
                    return value;
                }
                System.out.println("Please enter a number greater than zero.");
            } catch (NumberFormatException error) {
                System.out.println("Please enter a valid integer.");
            }
        }
    }

    static String readName(Scanner scanner, int index) {
        while (true) {
            System.out.print("Student #" + index + " name: ");
            String name = scanner.nextLine().trim();
            if (!name.isEmpty()) {
                return name;
            }
            System.out.println("Name cannot be empty.");
        }
    }

    static int readScore(Scanner scanner, String name) {
        while (true) {
            System.out.print(name + "'s score (0-100): ");
            String raw = scanner.nextLine().trim();
            try {
                int score = Integer.parseInt(raw);
                if (score >= 0 && score <= 100) {
                    return score;
                }
                System.out.println("Score must be between 0 and 100.");
            } catch (NumberFormatException error) {
                System.out.println("Please enter a valid integer score.");
            }
        }
    }

    public static void main(String[] args) {
        Locale.setDefault(Locale.US);
        Scanner scanner = new Scanner(System.in);
        int totalStudents = readPositiveCount(scanner);
        List<Student> students = new ArrayList<>();

        for (int index = 1; index <= totalStudents; index++) {
            String name = readName(scanner, index);
            int score = readScore(scanner, name);
            students.add(new Student(name, score));
        }

        int total = 0;
        int minimum = students.get(0).score();
        int maximum = students.get(0).score();

        System.out.println("Students:");
        for (Student student : students) {
            System.out.println("- " + student.name() + ": " + student.score());
            total += student.score();
            minimum = Math.min(minimum, student.score());
            maximum = Math.max(maximum, student.score());
        }

        double average = (double) total / students.size();
        System.out.printf("Average: %.2f%n", average);
        System.out.println("Minimum: " + minimum);
        System.out.println("Maximum: " + maximum);
    }
}
